_projects/automate_vulture_whitelist.md

```diff
@@ -0,0 +1,78 @@
+---
+name: "Automate Whitelist creation for vulture"
+collaborating_projects:
+  - coala
+  - vulture
+desc: "Automatically create whitelists and detect false positives for vulture."
+developers_involved:
+  - RJ722
+mentors:
+  - AbdealiJK
+  - jendrikseipp
+  - The-Compiler
+difficulty: medium
+status: in_progress
+initiatives:
+  - GSoC
+issues: []
+markdown: automate_vulture_whitelist.md
+requirements:
+  - "Familiarity with vulture and how whitelists work."
+tags:
+  - Vulture
+  - AST
+  - coverage
+---
+
+The aim of this project is to automatically create whitelists (to be utilized
+by [vulture](https://github.com/jendrikseipp/vulture)) for projects using sip
+files to communicate between C++ level API and Python programs, most
+significant of them being PyQt projects.
+
+#### IDEAS
+
+* Create a script which takes sip files as input and spits a whitelist with all
+  the virtual functions defined in the sip files. (Add Documentation and tests)
+* Use that script for generating a whitelist for popular projects using PyQt
+  (like [qutebrowser](https://github.com/qutebrowser/qutebrowser)).
+* Allow passing coverage.xml to vulture via `--make-whitelist` and let it find
+  the list of false-positive functions. (Add Documentation and tests as well)
+* Write a whitelist generator for shiboken, the binding
+  generator used for PySide2
+
+#### MILESTONES
+
+##### COMMUNITY BONDING
+* Discussion on whether to implement a new Tracer or to directly consume `xml`
+  output from coverage.py
+* Get familiar with coverage.xml
+* Take up small tasks from
+  [TODO](https://github.com/jendrikseipp/vulture/tree/master/TODO.rst)
+* [Follow up milestone](https://gitlab.com/coala/GSoC/GSoC-2018/milestones/28)
+
+##### CODING PHASE 1
+* The script for creating a whitelist from sip files should be implemented.
```

Comment:
`script` implies junky code.

You are creating a new repo, presumably a new package, and all code in it will be lovely.
And it will have a command line entry point.

Below you have this "script to create a whitelist for qutebrowser" , which is where the lower quality comes in -- in milestone 1 you are expecting that the very high quality code will be only guaranteed to support qutebrowser's needs.

I expect that this means that in milestone 1 you are supplying a PR to qutebrowser which replaces their whitelist generator with your own.

In the following milestones more use cases are added, and that will expand the scope of the codebase, dealing with additional scenarios not needed by qutebrowser.

Reply:
> I expect that this means that in milestone 1 you are supplying a PR to qutebrowser which replaces their whitelist generator with your own.

That [whitelist generator](https://github.com/qutebrowser/qutebrowser/blob/master/scripts/dev/run_vulture.py) would still need to stay, because it has many things in it which are qutebrowser specific. What it'd replace is the current [horrible heuristic](https://github.com/qutebrowser/qutebrowser/blob/v1.3.0/scripts/dev/run_vulture.py#L136) which ignores any functions which have a capital letter in them (because Qt/PyQt use `camelCase` for methods).

Reply:
I guess that means that the whitelist generator would need to also provide a API for the qutebrowser whitelist generator to use.

Reply:
No, we wouldn't need an API. The whitelist needs to be manually generated, saved and committed into the project. Even better would be when we start shipping it with vulture - That way whitelist would be automatically loaded whenever user imports Qt.